### PR TITLE
Update LRUCache.java

### PR DESCRIPTION
**Pull request by rishabhbohra3**

### CODE REVIEW

**Comment by rishabhbohra3** [comment]
Bug 1: Updates value after moving node to last
Bug 2: Fails to handle when entry is the head and only entry in the list
Bug 3: Default constructor might be initializing keys or values improperly
Bug 3: Constructor might not properly handle keys/values
Bug 4: Does not check if tail needs to be updated (tail might be the one evicted)
Bug 5: Does not evict entries if the new capacity is less than the current size

**Comment by rishabhbohra3** [comment]
## Code Review Agent Run Status

- **AI Based Review:** Successful
- **Static Analysis:** Partially Successful - Failed to execute static code analysis using fb_infer
## Code Review Overview

- **Summary:** The PR introduces a series of changes to the LRUCache implementation, focusing on refining the eviction logic, exception messages, and overall class structure. Key modifications include adjustments to capacity checks, exception message standardization, and the reorganization of the Entry class to streamline its definition. Additionally, the PR addresses potential NullPointerExceptions by ensuring proper checks before dereferencing objects.
- **Code change type:** Refactoring, Optimization
- **Unit tests added:** False
- **[Estimated effort to review ](https://docs.bito.ai/bito-dev-agents/ai-code-review-agent/faqs#what-is-estimated-effort-to-review-in-code-review-output)(1-5, lower is better):** 2




 <b> [>>See detailed code suggestions<<](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files) </b><br>The Bito AI Code Review Agent successfully reviewed 0 files and discovered no issues.
## High-level Feedback

 Overall, the PR significantly improves the LRUCache's reliability and maintainability. Future efforts should ensure that all changes, especially those affecting core logic, are accompanied by comprehensive unit tests to verify behavior under various scenarios and to prevent regressions.

**Comment by rishabhbohra3** [comment]
## Code Review Agent Run Status

- **AI Based Review:** Successful
- **Static Analysis:** Partially Successful - Failed to execute static code analysis using fb_infer
## Code Review Overview

- **Summary:** The PR introduces several changes and improvements to the LRUCache.java file, focusing on enhancing null safety, optimizing performance, and ensuring correct cache capacity checks. These changes aim to improve the reliability, performance, and maintainability of the cache implementation.
- **Code change type:** Optimisation, Bug Fix, Refactoring
- **Unit tests added:** False
- **[Estimated effort to review ](https://docs.bito.ai/bito-dev-agents/ai-code-review-agent/faqs#what-is-estimated-effort-to-review-in-code-review-output)(1-5, lower is better):** 2




 <b> [>>See detailed code suggestions<<](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files) </b><br>The Bito AI Code Review Agent successfully reviewed 1 file and discovered 3 issues. Please review these issues along with suggested fixes in the [Changed Files](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files).
## High-level Feedback

 Consider incorporating unit tests to cover the new changes, especially around null safety and cache capacity logic. Additionally, reviewing the exception handling strategy for empty cache scenarios could further refine the implementation. Aim for clear, maintainable code with a focus on performance and reliability.

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 1]
```diff
@@ -32,27 +21,27 @@ public LRUCache(int cap) {
...
         if (head == null) {
-            throw new RuntimeException("cache cannot be empty!");
+            throw new RuntimeException("Cache cannot be empty!");
         }
         Entry<K, V> evicted = head;
         head = evicted.getNextEntry();
-        head.setPreEntry(null);
+        if (head != null) {
+            head.setPreEntry(null);
+        }
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:36]: **Optimization Issue**: Missing null check before setting the previous entry to null can lead to NullPointerException. <br> **Fix**: Add a null check before setting the previous entry to ensure the head is not null. <br> **Code Suggestion**: 
 ```
 +        if (head != null) {
 +            head.setPreEntry(null);
 +        }
 ```

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 2]
```diff
@@ -32,27 +21,27 @@ public LRUCache(int cap) {
...
-            data.remove(evicted.getKey());
-        }
         this.cap = newCapacity;
+        
     }
 
     private Entry<K, V> evict() {
         if (head == null) {
-            throw new RuntimeException("cache cannot be empty!");
+            throw new RuntimeException("Cache cannot be empty!");
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:30]: **Performance Issue**: Throwing RuntimeException for a cache being empty could be an overuse of exceptions which might impact performance, especially if evict is called frequently in a scenario where the cache might be empty. <br> **Fix**: Consider using a specific checked exception or returning a special value to indicate an empty cache, which could be handled more gracefully. <br> **Code Suggestion**: 
 ```
 -            throw new RuntimeException("Cache cannot be empty!");
 +            // Consider returning null or a special value instead
 ```

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 3]
```diff
@@ -89,18 +79,15 @@ private void moveNodeToLast(Entry<K, V> entry) {
...
             moveNodeToLast(existingEntry);
+            existingEntry.setValue(value); 
             return;
         }
-        Entry<K, V> newEntry;
-        if (data.size() == cap) {
+        Entry<K, V> newEntry = new Entry<>(); 
+        if (data.size() >= cap) {
             newEntry = evict();
             data.remove(newEntry.getKey());
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:89]: **Scalability Issue**: Changing the condition from 'data.size() == cap' to 'data.size() >= cap' may lead to unnecessary eviction 
calls when the cache size is already below capacity. This could degrade the performance under high load, 
affecting scalability by performing redundant operations that could impact the throughput of cache operations. <br> **Fix**: Revert the condition to check for equality ('data.size() == cap') to avoid unnecessary evictions when the cache 
size is below the capacity. This ensures that evictions only occur when the cache is actually full, optimizing 
resource usage and maintaining high throughput under scale. <br> **Code Suggestion**: 
 ```
 -        if (data.size() >= cap) {
 +        if (data.size() == cap) {
 ```

**Comment by rishabhbohra3** [comment]
## Code Review Agent Run Status

- **AI Based Review:** Successful
- **Static Analysis:** Partially Successful - Failed to execute static code analysis using fb_infer
## Code Review Overview

- **Summary:** The PR introduces several key changes and improvements to the LRUCache.java class, focusing on enhancing performance, maintainability, and correctness. It refines capacity checking logic, error handling, and node eviction process, ensuring the cache operates efficiently under different scenarios. Additionally, it simplifies the Entry class structure by removing unnecessary constructors and methods, aiming for a leaner implementation.
- **Code change type:** Performance Improvement, Security, Refactoring
- **Unit tests added:** False
- **[Estimated effort to review ](https://docs.bito.ai/bito-dev-agents/ai-code-review-agent/faqs#what-is-estimated-effort-to-review-in-code-review-output)(1-5, lower is better):** 2




 <b> [>>See detailed code suggestions<<](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files) </b><br>The Bito AI Code Review Agent successfully reviewed 1 file and discovered 2 issues. Please review these issues along with suggested fixes in the [Changed Files](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files).
## High-level Feedback

 Focus on enhancing error handling to prevent potential security vulnerabilities by making messages more generic. Continue to streamline and optimize code structure for better maintainability and performance. Ensure all modifications are thoroughly tested, even if not explicitly mentioned in the feedback.

File: src/main/com/algos/src/datastructures/caches/LRUCache.java (hunk 2 above)
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:30]: **Security Issue**: Incorrect error handling could leak implementation details or system information, potentially aiding attackers in crafting further attacks. <br> **Fix**: Customize the error message to be less specific about the internal state or configuration of the cache system. <br> **Code Suggestion**: 
 ```
 -            throw new RuntimeException("Cache cannot be empty!");
 +            throw new RuntimeException("Cache operation failed due to empty state.");
 ```

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 4]
```diff
@@ -32,27 +21,27 @@ public LRUCache(int cap) {
...
+        }
         evicted.setNextEntry(null);
+        
         return evicted;
     }
 
     private void checkCapacity(int capacity) {
         if (capacity <= 0) {
-            throw new RuntimeException("capacity must greater than 0!");
+            throw new RuntimeException("Capacity must be greater than 0!");
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:44]: **Security Issue**: The error message for capacity check leaks details about the cache configuration, which could assist in attacks targeting the cache's capacity handling. <br> **Fix**: Reframe the error message to be generic and not disclose specific configuration constraints. <br> **Code Suggestion**: 
 ```
 -            throw new RuntimeException("Capacity must be greater than 0!");
 +            throw new RuntimeException("Invalid capacity value.");
 ```

**Comment by rishabhbohra3** [comment]
## Code Review Agent Run Status

- **AI Based Review:** Successful
- **Static Analysis:** Partially Successful - Failed to execute static code analysis using fb_infer
## Code Review Overview

- **Summary:** The PR introduces significant improvements to LRUCache.java, focusing on optimizing eviction logic and capacity management. It addresses high memory usage and inefficiency in eviction processes, suggesting a more efficient approach using a linked list for faster eviction and capacity checks. The changes aim to enhance performance, reduce CPU usage, and ensure the cache does not exceed its specified capacity.
- **Code change type:** Optimisation, Performance Improvement, Refactoring
- **Unit tests added:** False
- **[Estimated effort to review ](https://docs.bito.ai/bito-dev-agents/ai-code-review-agent/faqs#what-is-estimated-effort-to-review-in-code-review-output)(1-5, lower is better):** 2




 <b> [>>See detailed code suggestions<<](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files) </b><br>The Bito AI Code Review Agent successfully reviewed 1 file and discovered 3 issues. Please review these issues along with suggested fixes in the [Changed Files](https://github.com/rishabhbohra3/Algo-tutorial-test/pull/2/files).
## High-level Feedback

 Consider adding unit tests to cover the new eviction logic and capacity management features. Ensure that the refactoring does not introduce new bugs by thoroughly testing edge cases, especially around the eviction process and when the cache reaches its maximum capacity.

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 5]
```diff
@@ -32,27 +21,27 @@ public LRUCache(int cap) {
 
     private void setCapacity(int newCapacity) {
         checkCapacity(newCapacity);
-        for (int i = data.size(); i > newCapacity; i--) {
-            Entry<K, V> evicted = evict();
-            data.remove(evicted.getKey());
-        }
         this.cap = newCapacity;
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:24]: **Optimization Issue**: Removing the loop that adjusts the cache size during a capacity change may lead to excessive memory usage if the capacity is reduced significantly. <br> **Fix**: Reintroduce the loop to evict entries when the new capacity is less than the current size of the cache. This ensures that the cache does not hold more entries than the specified capacity. <br> **Code Suggestion**: 
 ```
 +        for (int i = data.size(); i > newCapacity; i--) {
 +            Entry<K, V> evicted = evict();
 +            data.remove(evicted.getKey());
 +        }
 ```

Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:24]: **Performance Issue**: Removing elements in a loop based on capacity check can be inefficient for large data sizes. This approach can lead to high CPU usage and slow down the performance due to constant resizing and rehashing of the HashMap. <br> **Fix**: Consider leveraging a linked list alongside the HashMap to efficiently remove the least recently used item without iterating over the map. This can significantly optimize the eviction process. <br> **Code Suggestion**: 
 ```
 -        for (int i = data.size(); i > newCapacity; i--) {
 -            Entry<K, V> evicted = evict();
 -            data.remove(evicted.getKey());
 -        }
 +        // Assume linked list implementation is used for ordering.
 +        while (data.size() > newCapacity) {
 +            Entry<K, V> evicted = evictLast();
 +            data.remove(evicted.getKey());
 +        }
 ```

File: src/main/com/algos/src/datastructures/caches/LRUCache.java [hunk 6]
```diff
@@ -32,27 +21,27 @@ public LRUCache(int cap) {
...
+            throw new RuntimeException("Cache cannot be empty!");
         }
         Entry<K, V> evicted = head;
         head = evicted.getNextEntry();
-        head.setPreEntry(null);
+        if (head != null) {
+            head.setPreEntry(null);
+        }
         evicted.setNextEntry(null);
+
```
Comment (rishabhbohra3) [inline src/main/com/algos/src/datastructures/caches/LRUCache.java:38]: **Scalability Issue**: Removing the loop that handles reducing the cache size to the new capacity upon resizing can lead to a cache size exceeding the specified capacity, negatively impacting memory scalability. <br> **Fix**: Reinstate the loop that iterates over the cache to evict entries until the cache size is equal to the new capacity, ensuring the cache does not exceed its maximum capacity. <br> **Code Suggestion**: 
 ```
 +        for (int i = data.size(); i > newCapacity; i--) {
 +            Entry<K, V> evicted = evict();
 +            data.remove(evicted.getKey());
 +        }
 ```
